### PR TITLE
Add support for double-width/double-height lines in conhost

### LINKED ISSUE #7865: Add support for double width/height lines
# Description of the new feature/enhancement

The VT100 (and later DEC terminals) had a bunch of operations that could alter the width and height of characters on a per-line basis: [`DECDWL`](https://vt100.net/docs/vt510-rm/DECDWL.html), [`DECDHL`](https://vt100.net/docs/vt510-rm/DECDHL.html), and [`DECSWL`](https://vt100.net/docs/vt510-rm/DECSWL.html). Despite their limitations, they're still quite useful for sprucing up a UI.

As an example, here's a UI for a chess app rendered with and without line attribute support:

image

# Proposed technical implementation details (optional)

I've been playing around with a POC of this for a while now, and I've come to realise there's actually a lot more work involved than you might first expect. Other than the basic rendering of characters, we need to handle text decoration, cursor rendering, selection marking, window resizing, mouse position detection, and every VT operation that depends on window boundaries will potentially need to take line attributes into account.

So before I spend more time on this, I wanted to make sure it's something you'd be happy to accept. I should say that I'm inclined to make it conhost-only to start with. I have also looked at the WT side of things, but that basically doubles the workload, and there are extra complications introduced by conpty that I don't have a good answer for.

### PR DESCRIPTION
**Pull request by j4james**
This PR adds support for the VT line rendition attributes, which allow
for double-width and double-height line renditions. These renditions are
enabled with the `DECDWL` (double-width line) and `DECDHL`
(double-height line) escape sequences. Both reset to the default
rendition with the `DECSWL` (single-width line) escape sequence. For now
this functionality is only supported by the GDI renderer in conhost.

There are a lot of changes, so this is just a general overview of the
main areas affected.

Previously it was safe to assume that the screen had a fixed width, at
least for a given point in time. But now we need to deal with the
possibility of different lines have different widths, so all the
functions that are constrained by the right border (text wrapping,
cursor movement operations, and sequences like `EL` and `ICH`) now need
to lookup the width of the active line in order to behave correctly.

Similarly it used to be safe to assume that buffer and screen
coordinates were the same thing, but that is no longer true. Lots of
places now need to translate back and forth between coordinate systems
dependent on the line rendition. This includes clipboard handling, the
conhost color selection and search, accessibility location tracking and
screen reading, IME editor positioning, "snapping" the viewport, and of
course all the rendering calculations.

For the rendering itself, I've had to introduce a new
`PrepareLineTransform` method that the render engines can use to setup
the necessary transform matrix for a given line rendition. This is also
now used to handle the horizontal viewport offset, since that could no
longer be achieved just by changing the target coordinates (on a double
width line, the viewport offset may be halfway through a character).

I've also had to change the renderer's existing `InvalidateCursor`
method to take a `SMALL_RECT` rather than a `COORD`, to allow for the
cursor being a variable width. Technically this was already a problem,
because the cursor could occupy two screen cells when over a
double-width character, but now it can be anything between one and four
screen cells (e.g. a double-width character on the double-width line).

In terms of architectural changes, there is now a new `lineRendition`
field in the `ROW` class that keeps track of the line rendition for each
row, and several new methods in the `ROW` and `TextBuffer` classes for
manipulating that state. This includes a few helper methods for handling
the various issues discussed above, e.g. position clamping and
translating between coordinate systems.

## Validation Steps Performed

I've manually confirmed all the double-width and double-height tests in
_Vttest_ are now working as expected, and the _VT100 Torture Test_ now
renders correctly (at least the line rendition aspects). I've also got
my own test scripts that check many of the line rendition boundary cases
and have confirmed that those are now passing.

I've manually tested as many areas of the conhost UI that I could think
of, that might be affected by line rendition, including things like
searching, selection, copying, and color highlighting. For
accessibility, I've confirmed that the _Magnifier_ and _Narrator_
correctly handle double-width lines. And I've also tested the Japanese
IME, which while not perfect, is at least useable.

Closes #7865

### CODE REVIEW

**Comment by j4james** [comment]
I've tried to keep this as simple and efficient as I could, but it touches on almost every part of the system, so it seems inevitable that it's going to have some impact on performance. And given that there isn't much demand for this functionality, I'd completely understand if you decide not to merge this.

That said, here are a couple of screenshots so you can see how well it handles _VT Test_ and the _VT100 Torture Test_.

image

image

**Comment by DHowett** [comment]
_wow_

**Comment by miniksa** [comment]
> I've tried to keep this as simple and efficient as I could, but it touches on almost every part of the system, so it seems inevitable that it's going to have some impact on performance. And given that there isn't much demand for this functionality, I'd completely understand if you decide not to merge this.

James... with high quality contributions like this, it would be a crime for us to not merge it at some point. I'll even catch your back with the performance implications and writing out the DX version of the rendering if I must. It just might take us a little bit of time to digest the whole thing before we can take it.

Once again, thank you so much. I hope to read through it soon.

File: src/terminal/adapter/adaptDispatch.cpp [hunk 1]
```diff
@@ -622,6 +621,23 @@ bool AdaptDispatch::EraseInDisplay(const DispatchTypes::EraseType eraseType)
...
+        // When erasing the display, every line that is erased in full should be
+        // reset to single width. When erasing to the end, this could include
+        // the current line, if the cursor is in the first column. When erasing
+        // from the beginning, though, the current line would never be included,
+        // because the cursor could never be in the rightmost column (assuming
+        // the line is double width).
+        if (eraseType == DispatchTypes::EraseType::FromBeginning)
+        {
+            const auto endRow = csbiex.dwCursorPosition.Y;
+            _pConApi->PrivateResetLineRenditionRange(csbiex.srWindow.Top, endRow);
```
Comment (miniksa) [inline src/terminal/adapter/adaptDispatch.cpp:633]: Why reset the line rendition separately here instead of inside of `DoSrvPrivateFillRegion` or even further down like in every consumer of `screenInfo.Write`/`TextBuffer::WriteLine` much like how line wraps are added/removed/left alone in certain common cases like rectangular fills?

**Review by miniksa — CHANGES_REQUESTED** [review]
Looking pretty good. Just a couple questions and such.

I'll pull a perf trace on this branch and master with GDI to see if there's any noticeable detriment to this in the common case where it's unused.

**Comment by miniksa** [comment]
This is the only thing I noticed in writing `big.txt` between `main` and this branch. Note that big.txt doesn't output anything but standard text in standard lines.

image

It looks like we're calling `Set/ModifyWorldTransform` to GDI even when there isn't really anything to talk about. Unfortunately when I expand it, it looks like this goes all the way into kernel to resolve that there's nothing to change. So if we could not call it when there's nothing to be done on our side, it would probably save us something.

image

File: src/buffer/out/LineRendition.hpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+{
+    SingleWidth,
+    DoubleWidth,
+    DoubleHeightTop,
+    DoubleHeightBottom
+};
+
+constexpr SMALL_RECT ScreenToBufferLine(const SMALL_RECT& line, const LineRendition lineRendition)
+{
+    const auto scale = lineRendition == LineRendition::SingleWidth ? 0 : 1;
```
[the remark this answers is not in the document]
Comment (j4james) [reply src/buffer/out/LineRendition.hpp:26]: Yeah, there's probably an element of premature optimization here. I was less concerned about speed though - I honestly wouldn't know if an imul or idiv is significantly worse than a shift on a modern processor. But I was horrified by the amount of code I was introducing in my initial implementation, and the shifts were definitely an improvement in that respect.

The current version is actually a bit of a compromise - the optimizer doesn't handle inlining of these functions very well - but I was also trying to keep some of the messy details hidden in the helper functions, and not bloat the rest of the source too much.

There was also another reason I preferred the shifts - at least for division. If the values are negative, then a div doesn't round in the right direction, whereas a shift does. I don't _think_ we need to support negative numbers in the current usage, but it felt safer knowing that these functions would do the right thing if that ever did occur.

I agree this could definitely do with some commenting though.

File: src/buffer/out/textBuffer.cpp [hunk 3]
```diff
@@ -2057,9 +2133,19 @@ HRESULT TextBuffer::Reflow(TextBuffer& oldBuffer,
     {
         // Fetch the row and its "right" which is the last printable character.
         const ROW& row = oldBuffer.GetRowByOffset(iOldRow);
+        const short cOldColsTotal = oldBuffer.GetLineWidth(iOldRow);
         const CharRow& charRow = row.GetCharRow();
         short iRight = gsl::narrow_cast<short>(charRow.MeasureRight());
 
+        // If we're starting a new row, try and preserve the line rendition
```
[the remark this answers is not in the document]
Comment (j4james) [reply src/buffer/out/textBuffer.cpp:2140]: > Presumably if you have a double height row, shrink the window, it'll run through reflow and now you'll have... 2 rows that are double height. Right?

No. A double height row is basically two lines with identical text - one rendering the double height top, and the other rendering the bottom. Once the top portion wraps, it's just going to split the double height rendition in half with a normal size line in between. No matter what line attribute you assign to the wrapped portion, it's not going to make much sense, so I don't even try to do anything fancy. But if you regrow the window and it unwraps again, it should at least go back to normal.

I kind of went with a minimalist solution here. Basically it should work reasonably for the simple cases, but once the content wraps, I don't think there are any good solutions - these attributes were just never expected to be used that way.

File: src/buffer/out/textBuffer.hpp [hunk 4]
```diff
@@ -141,7 +151,7 @@ class TextBuffer final
     bool MoveToNextGlyph(til::point& pos, bool allowBottomExclusive = false) const;
     bool MoveToPreviousGlyph(til::point& pos) const;
 
-    const std::vector<SMALL_RECT> GetTextRects(COORD start, COORD end, bool blockSelection = false) const;
+    const std::vector<SMALL_RECT> GetTextRects(COORD start, COORD end, bool blockSelection = false, bool bufferCoordinates = false) const;
```
[the remark this answers is not in the document]
Comment (j4james) [reply src/buffer/out/textBuffer.hpp:154]: Outside of the unit tests, it looks like there are only two that are false (and three that are true). Would it perhaps be better to get rid of the default value, and always just set the `bufferCoordinates` parameter explicitly? We can get rid of the `blockSelection` default too, since that's already always being set explicitly (except maybe some of the unit tests, but I think there are only a handful of those).

File: src/renderer/base/renderer.cpp [hunk 5]
```diff
@@ -950,18 +984,34 @@ void Renderer::_PaintBufferOutputGridLineHelper(_In_ IRenderEngine* const pEngin
         // bottom of the viewport, the space that's not quite a full line in
         // height. Since we don't draw that text, we shouldn't draw the cursor
         // there either.
-        Viewport view = _pData->GetViewport();
-        if (view.IsInBounds(coordCursor))
+
+        // The cursor is never rendered as double height, so we don't care about
+        // the exact line rendition - only whether it's double width or not.
+        const auto doubleWidth = _pData->GetTextBuffer().IsDoubleWidthLine(coordCursor.Y);
+        const auto lineRendition = doubleWidth ? LineRendition::DoubleWidth : LineRendition::SingleWidth;
```
[the remark this answers is not in the document]
Comment (j4james) [reply src/renderer/base/renderer.cpp:991]: As explained above, a double height line is actually two lines with identical content. So the cursor could be on either the top half or the bottom half of that line, and would only cover the height of the portion that it's on. Technically both halves may not even be there - you could quite easily have the bottom half of a double height line with a single height line above it.

That said, I don't know for a fact that the original hardware terminals worked that way. It just seemed the most reasonable approach to take, and I haven't come across any terminal emulators that do things differently.

File: src/renderer/gdi/paint.cpp [hunk 6]
```diff
@@ -41,6 +41,9 @@ using namespace Microsoft::Console::Render;
     _psInvalidData.hdc = GetDC(_hwndTargetWindow);
     RETURN_HR_IF_NULL(E_FAIL, _psInvalidData.hdc);
 
+    // We need the advanced graphics mode in order to set a transform.
+    SetGraphicsMode(_psInvalidData.hdc, GM_ADVANCED);
```
[the remark this answers is not in the document]
Comment (j4james) [reply src/renderer/gdi/paint.cpp:45]: > Does it impact the fidelity of the scaled text at all (is extra aliasing visible) or does it work perfectly as if you had set the font size larger?

It looks pretty smooth to me. It's definitely not just pixel-doubling. However, I should mention that I've noticed block characters not always lining up perfectly when double sized, but I'm not sure that's necessarily the fault of the transforms, and I didn't think it was worth worrying about too much until someone complained.

> it might cause all sorts of GDI stuff to take a different codepath that is slower in the common case

Yeah, I did wonder about that too. I'm not sure you can just turn it on and off as you need it though. And note that the renderer is now using transforms for the viewport offset too, so it's not just the double-size renditions that need it. But if it it's a significant problem, I can try and see what it would take to enable it on demand - I suspect it would be messy though.

File: src/renderer/uia/UiaRenderer.hpp [hunk 7]
```diff
@@ -86,6 +86,6 @@ namespace Microsoft::Console::Render
         Microsoft::Console::Types::IUiaEventDispatcher* _dispatcher;
 
         std::vector<SMALL_RECT> _prevSelection;
-        til::point _prevCursorPos;
+        SMALL_RECT _prevCursorRegion;
```
[the remark this answers is not in the document]
Comment (j4james) [reply src/renderer/uia/UiaRenderer.hpp:89]: Yeah, I did go through a few different options here, but it just made everything so much easier to stick with the `SMALL_RECT` that was being used everywhere else.

File: src/terminal/adapter/adaptDispatch.cpp (hunk 1 above)
Comment (j4james) [reply src/terminal/adapter/adaptDispatch.cpp:633]: I did consider doing something like that, but not all rectangular fills are meant to reset the line rendition. For example you could clear a double-width line with `EL` or `ECH`, and that would result in the full width being filled, but shouldn't reset the rendition.

**Comment by j4james** [comment]
> It looks like we're calling `Set/ModifyWorldTransform` to GDI even when there isn't really anything to talk about.

I'm really glad you spotted that. I actually had checks in place to avoid setting and resetting the transform unnecessarily in the `PrepareLineTransform` and `ResetLineTransform` methods, but I got caught out by the fact that -0.0 and 0.0 aren't equal.

And I just realised there's also the special case handling for the inverted cursor, which has to "uninvert" itself when scrolling, and that requires setting a transform first. So that's also going to need a check to avoid the transform when it's just the identity matrix.

File: src/terminal/adapter/adaptDispatch.cpp (hunk 1 above)
Comment (miniksa) [reply src/terminal/adapter/adaptDispatch.cpp:633]: Fair enough.

File: src/buffer/out/textBuffer.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (miniksa) [reply src/buffer/out/textBuffer.hpp:154]: Sure. I like making it explicit and cleaning that up.

File: src/buffer/out/textBuffer.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (miniksa) [reply src/buffer/out/textBuffer.cpp:2140]: Ah that's fine. Okay. Thanks for clarifying.

File: src/renderer/base/renderer.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (miniksa) [reply src/renderer/base/renderer.cpp:991]: Alright.

File: src/renderer/gdi/paint.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (miniksa) [reply src/renderer/gdi/paint.cpp:45]: Nah based on the wpr... It didn't have a big impact just for setting the mode. So ignore this one.

**Comment by j4james** [comment]
OK. I've added the requested comments, made the `GetTextRects` parameters explicit, and improved the checks to avoid unnecessary transform changes. You shouldn't be seeing any `Set/ModifyWorldTransform` calls now if you're just rendering regular single-width text, with no horizontal viewport offset.

**Review by miniksa — APPROVED** [review]
Works for me. Thanks, @j4james.

**Review by hafixo — APPROVED** [review]
Merge pull requested #8864 and close #8867

**Review by hafixo — CHANGES_REQUESTED** [review]
Close

**Comment by zadjii-msft** [comment]
@hafixo Please stop commenting on unrelated PRs without actionable feedback.

**Comment by hafixo** [comment]
> @hafixo Tolong berhenti memberi komen mengenai PR yang tidak berkaitan tanpa maklum balas yang boleh diambil tindakan.

Im sorry..,im hit more button because my page so lagging to view code... apologize me

**Comment by DHowett** [comment]
Sorry, @j4james, this is _all_ on me. I've been travelling and not had a chance to look over this in-depth. I'll clear the stale state and get it in this week. :smile:

Thanks again for excellent work.

**Comment by j4james** [comment]
@DHowett Take your time. I've been messing about with this for months, so I'm in no hurry to get it merged.

File: src/host/getset.cpp [hunk 8]
```diff
@@ -1410,6 +1415,10 @@ void DoSrvPrivateAllowCursorBlinking(SCREEN_INFORMATION& screenInfo, const bool
     {
         cursorPosition.X = 0;
     }
+    else
+    {
+        cursorPosition = textBuffer.ClampPositionWithinLine(cursorPosition);
```
Comment (DHowett) [inline src/host/getset.cpp:1420]: This is for when a LF moves you into a non-single-width line, right?

Assuming there's no deferred EOL breakage we're concerned with here since this _is_ a cursor movement.

**Review by DHowett — CHANGES_REQUESTED** [review]
I'm at 34/50 files, and going to call it a night (it's 11:30 PM here). I checked it out, merged main and built it. It's _so_ freakin' cool! I've got a couple notes from using it.

* Double-click word selection seems to treat the click and expansion coordinates as _buffer_ instead of _screen_ (resulting in the wrong range being selected)
* I encountered a crash resizing the console down (horizontally) after putting a normally double-width glyph into a double-width line²
   * `printf "\e#3Hello World\u30ab\n\e#4Hello World\u30ab\n"`
* I don't expect this to work, but MAN this makes COOKED_READ angry. It stops wrapping (even at buffer width). I also don't expect for you to fix that!

² very surprised to see a 4-cell-wide glyph work properly, wow!

Blocking for the crash, and I guess a little bit because I didn't get all the way through it.

**Comment by j4james** [comment]
> * Double-click word selection seems to treat the click and expansion coordinates as _buffer_ instead of _screen_ (resulting in the wrong range being selected)

I thought I might miss a few of these operations. But that should be an one easy to fix. I think it just needs a couple of `ScreenToBuffer` and `BufferToScreen` calls in the `SCREEN_INFORMATION::GetWordBoundary` method.

> * I encountered a crash resizing the console down (horizontally) after putting a normally double-width glyph into a double-width line²

Oof. That's bad. Thanks for catching that. It looks like the `TextBuffer::_GetPreviousFromCursor` method needs to be adjusted to take the current line width into account rather than assuming the full buffer width. If that's all it is, that should be an easy fix too. 

> * I don't expect this to work, but MAN this makes COOKED_READ angry. It stops wrapping (even at buffer width). I also don't expect for you to fix that!

Yeah, I wouldn't have expected that work either. I assume if you're mixing VT sequences with console APIs then weird things are going to happen. I suspect this isn't the only escape sequence that'll freak out COOKED_READ.

File: src/host/getset.cpp (hunk 8 above)
Comment (j4james) [reply src/host/getset.cpp:1420]: Correct. The delayed EOL flag is going to be reset as soon as we set the cursor position, so I don't think this should break anything in that regard.

File: src/host/screenInfo.cpp [hunk 9]
```diff
@@ -2464,6 +2469,12 @@ std::pair<COORD, COORD> SCREEN_INFORMATION::GetWordBoundary(const COORD position
             }
         }
     }
+
+    // The calculated range is in buffer coordinates, but the caller is
+    // expecting screen offsets, so we have to convert these back again.
```
Comment (DHowett) [inline src/host/screenInfo.cpp:2474]: This is thanks to you adding a `false` (for useBufferCoordinates) to the text rect computation in Selection, right?

I couldn't immediately evaluate this as true, but I did look at consumers in Selection of the endpoints that are populated through this function and trust it well enough.

Comment (j4james) [reply src/host/screenInfo.cpp:2474]: The only place `GetWordBoundary` is called is from the window mouse event handler. It is passed the mouse position in screen coordinates, and then expects to receive the calculated word boundary in the same coordinate system. See here:
https://github.com/microsoft/terminal/blob/b1e1c7cdf4efa1ca4e0a495f6b90fd83ca8b5633/src/interactivity/win32/windowio.cpp#L803-L806

Comment (DHowett) [reply src/host/screenInfo.cpp:2474]: Right -- and that's fed into Selection, which (as far as I can tell) only uses it for updating the renderer's invalidation range. I couldn't determine if anyone was using the stuff in Selection in buffer coords. :smile:

**Comment by j4james** [comment]
When testing the fix for the crash which occurred when reflowing a double width line with double width characters, I noticed that the double width characters could end up split in half as they were forced to wrap onto another line. I've since fixed that in commit 55c93dc7e944e5c3e9d4a4e51cf7ac07b5274952.

**Comment by DHowett** [comment]
This will regrettably need a second merge from main (or rebase), but I'm ready to set automerge and click that button. :smile:

File: src/terminal/adapter/adaptDispatch.cpp [hunk 10]
```diff
@@ -1931,6 +1959,8 @@ bool AdaptDispatch::ScreenAlignmentPattern()
         auto fillPosition = COORD{ 0, csbiex.srWindow.Top };
         const auto fillLength = (csbiex.srWindow.Bottom - csbiex.srWindow.Top) * csbiex.dwSize.X;
         success = _pConApi->PrivateFillRegion(fillPosition, fillLength, L'E', false);
+        // Reset the line rendition for all of these rows.
+        success = success && _pConApi->PrivateResetLineRenditionRange(csbiex.srWindow.Top, csbiex.srWindow.Bottom);
```
Comment (DHowett) [inline src/terminal/adapter/adaptDispatch.cpp:1963]: Huh, we fill *after* we reset the rendition? I don't totally understand why, but I trust your homework here.

**Review by DHowett — APPROVED** [review]

File: src/terminal/adapter/adaptDispatch.cpp (hunk 10 above)
Comment (DHowett) [reply src/terminal/adapter/adaptDispatch.cpp:1963]: erm we fill *before*.

Comment (j4james) [reply src/terminal/adapter/adaptDispatch.cpp:1963]: I don't think it matters. The fill operation doesn't care about line renditions, so I suspect it would work in either order. I do know that this sequence works correctly as is - it's one of my manual test cases.

Comment (j4james) [reply src/terminal/adapter/adaptDispatch.cpp:1963]: I've just double checked this and the order doesn't make any difference. In both case the lines are reset to single width and the screen is completely filled with Es.

File: src/host/screenInfo.cpp (hunk 9 above)
Comment (j4james) [reply src/host/screenInfo.cpp:2474]: It is a bit confusing. Different parts of the `Selection` class use different coordinate systems. For example the `ColorSelection` methods expect buffer coordinates, and the `GetSelectionRects` method returns buffer coordinates, but the internal anchor positions are stored in screen coordinates. This was a bit of a nightmare to sort out, and although I tried my best to catch as many of these cases as I could, it wouldn't be a huge surprise to me if there were others that I missed.

**Comment by j4james** [comment]
FYI, I'm going to bed now, so if there is many more merging or changes required it'll have to wait until tomorrow.

**Comment by DHowett** [comment]
Beautiful work with great attention to detail, as always. Thanks again @j4james!

**Comment by christianparpart** [comment]
What the... I am very glad to see you having implemented this. Good work!

**Comment by DHowett** [comment]
This is available in conhost as of [Windows Insider Preview build 21337](https://blogs.windows.com/windows-insider/2021/03/17/announcing-windows-10-insider-preview-build-21337/). Congrats! :smile:

**Review by hafixo — COMMENTED** [review]
#8664
Reviews and change
